### PR TITLE
Subset of primitive types support for serialization [API-1179]

### PR DESCRIPTION
**Pull request by utku-caglayan**
Introduce basic primitive types for serialization for map keys and values for types

Integers (8, 16, 32, 64)
Boolean
Floats (32, 64)

### CODE REVIEW

File: types/mapcmd/map.go [hunk 1]
```diff
@@ -109,17 +109,22 @@ func decorateCommandWithMapNameFlags(cmd *cobra.Command, mapName *string) {
 	cmd.MarkFlagRequired("name")
 }
 
-func decorateCommandWithKeyFlags(cmd *cobra.Command, mapKey *string) {
-	cmd.Flags().StringVarP(mapKey, "key", "k", "", "key of the map")
+func decorateCommandWithKeyFlags(cmd *cobra.Command, mapKey, mapKeyType *string) {
+	flags := cmd.Flags()
+	flags.StringVarP(mapKey, "key", "k", "", "key of the map")
+	flags.StringVarP(mapKeyType, "key-type", "", "string", "type of the key, one of: string, json, int(8,16,32,64), float(32,64)")
```
Comment (srknzl) [inline types/mapcmd/map.go:115]: Maybe it is better to list all of them like int8, int16 etc. What do you think? But regardless of it, looks like it does not fit to the prompt: 
image

Can we make it fit via displaying it in multiple lines etc?

File: types/mapcmd/map.go [hunk 2]
```diff
@@ -109,17 +109,22 @@ func decorateCommandWithMapNameFlags(cmd *cobra.Command, mapName *string) {
...
 	cmd.MarkFlagRequired("key")
+	cmd.RegisterFlagCompletionFunc("key-type", func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
```
Comment (srknzl) [inline types/mapcmd/map.go:117]: If this should autocomplete in the prompt, I was not able to trigger it:

image

Is it because I build from source code?

File: internal/conversion.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+		intermediateInt   int64
+		intermediateFloat float64
+	)
+	switch valueType {
+	case String:
+		convertedValue = value
+	case Boolean:
+		convertedValue, err = strconv.ParseBool(value)
+	case JSON:
+		convertedValue = serialization.JSON(value)
```
Comment (yuce) [inline internal/conversion.go:50]: I think it would be nice to check whether the given JSON is valid by trying to unmarshal it.

File: internal/conversion.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+		convertedValue = float32(intermediateFloat)
+	case Float64:
+		convertedValue, err = strconv.ParseFloat(value, 64)
+	default:
+		err = fmt.Errorf("unknown type, provide one of %v", SupportedTypes)
+	}
+	if numErr, ok := err.(*strconv.NumError); ok {
+		if numErr.Err == strconv.ErrSyntax {
+			err = fmt.Errorf(`can not convert "%s" to %s, unknown syntax`, value, valueType)
+		} else if numErr.Err == strconv.ErrRange {
```
Comment (yuce) [inline internal/conversion.go:73]: You can use `errors.Is(err, strconv.ErrSyntax)` and `errors.Is(err, strconv.ErrRange)` instead.

File: internal/conversion.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+		convertedValue = int32(intermediateInt)
+	case Int64:
+		convertedValue, err = strconv.ParseInt(value, 10, 64)
+	case Float32:
+		intermediateFloat, err = strconv.ParseFloat(value, 32)
+		convertedValue = float32(intermediateFloat)
+	case Float64:
+		convertedValue, err = strconv.ParseFloat(value, 64)
+	default:
+		err = fmt.Errorf("unknown type, provide one of %v", SupportedTypes)
```
Comment (yuce) [inline internal/conversion.go:68]: How about something like: `strings.Join(SupportedTypes, ", ")` instead of `%v`?

File: internal/conversion.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+func ConvertString(value, valueType string) (interface{}, error) {
+	var (
+		convertedValue interface{}
+		err            error
+	)
+	var (
+		intermediateInt   int64
+		intermediateFloat float64
```
Comment (yuce) [inline internal/conversion.go:42]: These variable names are too long. How about just `i` and `f`?

File: internal/conversion.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	Int16,
+	Int32,
+	Int64,
+	Float32,
+	Float64,
+}
+
+func ConvertString(value, valueType string) (interface{}, error) {
+	var (
+		convertedValue interface{}
```
Comment (yuce) [inline internal/conversion.go:37]: `convertedValue` is a bit long. how about just `cv`?

File: types/mapcmd/map.go [hunk 8]
```diff
@@ -74,7 +74,7 @@ func getMap(ctx context.Context, clientConfig *hazelcast.Config, mapName string)
 	return
 }
 
-const MapUseExample = "map use m1          # sets the default map name to m1 unless set explicitly"
+const MapUseExample = "map use m1    # sets the default map name to m1 unless set explicitly"
```
Comment (yuce) [inline types/mapcmd/map.go:77]: Why is this change necessary?

**Review by yuce — CHANGES_REQUESTED** [review]

File: internal/conversion.go (hunk 6 above)
Comment (utku-caglayan) [reply internal/conversion.go:42]: I was going to name them tmp, but figured you would not be happy with that 😬

File: internal/conversion.go (hunk 3 above)
Comment (utku-caglayan) [reply internal/conversion.go]: JSON validation on https://github.com/hazelcast/hazelcast-commandline-client/pull/71/commits/e1b36476184397d20088747e12f583a767f096b1

File: internal/conversion.go (hunk 5 above)
Comment (utku-caglayan) [reply internal/conversion.go]: Changed https://github.com/hazelcast/hazelcast-commandline-client/pull/71/commits/e1b36476184397d20088747e12f583a767f096b1

File: internal/conversion.go (hunk 7 above)
Comment (utku-caglayan) [reply internal/conversion.go]: Renamed https://github.com/hazelcast/hazelcast-commandline-client/pull/71/commits/e1b36476184397d20088747e12f583a767f096b1

File: types/mapcmd/map.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply types/mapcmd/map.go]: Fixed https://github.com/hazelcast/hazelcast-commandline-client/pull/71/commits/e1b36476184397d20088747e12f583a767f096b1

File: types/mapcmd/map.go (hunk 8 above)
Comment (utku-caglayan) [reply types/mapcmd/map.go:78]: On a previous PR we decided to remove tabs and use spaces

**Review by yuce — APPROVED** [review]
Looks good!

**Review by srknzl — APPROVED** [review]
